### ISSUE TITLE
Override in config.*.yaml of false/null/empty values does not work anymore in v1.20.0

### ISSUE THREAD

**Issue by cmuench** [drud/ddev#4079]
### Is there an existing issue for this?

- [X] I have searched the existing issues

https://gist.github.com/cmuench/6c633b58d33ec9fcf802b6bc2dda551e

### Current Behavior

I have the NFS mount enabled in the projects `.ddev/config.yaml` file.
In a `.ddev/config.local.yaml` the setting is set to `false`.

If I start the project then the NFS mount is used.

A `ddev debug configyaml | grep nfs` prints `nfs_mount_enabled: true`.

I also tested with other values likes "timezone" or "composer_version". That values are overridden as expected.
My suggestion is that the NFS mount config has a different handling since v1.20.0

### Expected Behavior

If value in config.local.yaml file should override.
NFS mount should *not* be used in the project.

### Steps To Reproduce

1. Create project
2. Set "nfs_mount_enabled" to true in config.yaml
3. Create a config.local.yaml and set "nfs_mount_enabled" to false
4. Start the project
5. NFS mount is used

### Anything else?

It's currently only nfs_mount_enabled affected.

**Comment by rfay** [comment]
Thanks, I'll take a look. There's no new handling, but I guess maybe this being a bool could be different? 

I didn't think anybody was using nfs any more in the era of mutagen. It sounds like you're not using it either...

Note that enabling mutagen also wins over nfs_mount_enabled.

**Comment by cmuench** [comment]
@rfay Mutagen is nice but it duplicates all the content. If you work with several projects then you want to save space and redurce the startup time.
In our project we have Macbooks and the NFS is enabled by default per project. In my Linux environment I disable the NFS mount then.

IMHO It's not super critical but a different behavior.

**Comment by rfay** [comment]
(Note that since v1.19.0, only the code is duplicated, not the uploaded files, so it's a far smaller amount of duplication now.)

**Comment by cmuench** [comment]
Right but it's still a lot of data to sync if it's the Magento code base. :-)

**Comment by rfay** [comment]
For interest's sake I checked, and the mutagen volume from a magento2 quickstart contains 821MB. 

Trying to figure out what's going on with nfs_mount_enabled.

**Comment by rfay** [comment]
This uses mergo.Merge(), https://github.com/imdario/mergo

```
// Merge will fill any empty for value type attributes on the dst struct using corresponding
// src attributes if they themselves are not empty. dst and src must be valid same-type structs
// and dst must be a pointer to struct.
// It won't merge unexported (private) fields and will do recursively any exported field.
func Merge(dst, src interface{}, opts ...func(*Config)) error {
	return merge(dst, src, opts...)
}
```

"Merge will fill any empty for value type attributes on the dst struct using corresponding src attributes if they themselves are not empty". 

`false` is considered nil by golang bool type. So I guess that's the cause of the problem. And I think `nfs_mount_enabled` is the only config element that is a bool that would be relevant to anybody.

**Comment by rfay** [comment]
I'm going to have to think about this. @torenware please think about it as well. It turns out that a boolean `false` is the same as "not set" and is the default. So it will take special code that says "is this different from the existing config", and even that may be hard.

**Comment by cmuench** [comment]
Maybe we can internally work with a "Boolish-String" and parse them via `strconv.ParseBool()`? Yes, it's a kind of hack. :)

**Comment by rfay** [comment]
I guess this will also happen with empty strings, probably overwriting 
* `docroot: web` with `docroot: ""` 
* `mutagen_enabled` 
* `use_dns_when_possible`
*  `composer_version: ""`
* `xdebug_enabled`
* `xhprof_enabled`
* `timezone: ""`
* `composer_root: ""`
* working_dir
* upload_dir
* fail_on_hook_fail
* disable_settings_management
* no_project_mount
* bind_all_interfaces


Essentially, every case where you're trying to override "something" with `nil` value.  Most likely to happen with bools, but empty string overrides are most likely implicated. I don't think there are any likely ints.

**Comment by rfay** [comment]
I don't think `strconv.ParseBool()` would accomplish anything, because the loaded `app` we're merging will always have the false/empty value, even when you haven't overridden it.

**Comment by torenware** [comment]
> I'm going to have to think about this. @torenware please think about it as well. It turns out that a boolean `false` is the same as "not set" and is the default. So it will take special code that says "is this different from the existing config", and even that may be hard.

The joys of Golang "empty" values? I'll give it a think.

**Comment by rfay** [comment]
And of course, not just binary, but any feature whose default is a nil-value.

**Comment by torenware** [comment]
One feature of Mergo that may come in handy for this: it turns out to be extremely simple to convert a struct into a map[string]any.  Run this by a type switch, and you have very easy compare of the scalar fields. Not sure if you can make a round trip with this, but if it's possible, then the code to handle the compares becomes very simple.

**Comment by rfay** [comment]
A proposal for a technique:
* https://github.com/drud/ddev/issues/4100

**Comment by torenware** [comment]
> A proposal for a technique:
> 
>     * [Provide feature flag for overriding config.*.yaml instead of merging #4100](https://github.com/drud/ddev/issues/4100)

He mentions yq, which is a great idea, since yq is small, and allows us to test for a particular item in the yaml.  Suppose we defined a special key/value pair for `config_override`, where we behaved as if the default was `false`.  Then for shell files, we could test if a yaml file contained `config_override: true` and special case its processing. Note this is easy enough to do Go directly, since we can parse in a file as map[string]any and do the same test.  We'd then handle a config.*.yaml file according to whether the setting was present or not, using the current merging code if present, and doing a straight override if not.

This would remove any need for a special delete-this-setting syntax. The file would be all merge or all replace.  Which I think is a simpler dev experience.

**Comment by rfay** [comment]
I'd love it if you could all review and test the PR for this, https://github.com/drud/ddev/pull/4118#issuecomment-1214230090

**Comment by rfay** [comment]
Hi @cooldrupal  - This was already fixed in v1.21.1, see docs, https://ddev.readthedocs.io/en/latest/users/extend/customization-extendibility/#extending-configyaml-with-custom-configyaml-files

You can use the `override_config: true` when you need it.

**Comment by cooldrupal** [comment]
@rfay , many thanks, it works with `override_config` parameter.
